### DISCUSSION TITLE
Proposal: Update Orders' Total and Tax Handling

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#6121] [Ideas]
Zen Cart currently uses _distributed_ approach to the calculation of an order's total-value and tax.  The tax on each element of an order, whether the products themselves or order-total additions, is calculated separately as the associated element is added to the order.

This approach can lead to
1. Penny-off discrepancies between the values stored in the order-object, i.e. `$order->info['total']` and `$order->info['tax']`, and those stored in the associated order-totals, i.e. `$ot_total['value']` and `$ot_tax['value']`.  See #2735 for an example.
2. Over-complication of calculations for order-related deductions, e.g. coupons or reward-points, and order editing, e.g. Edit Orders.
3. Lack of an audit trail, since there is no history recorded for the tax-related calculations.
4. Increased debug time, if any calculations "go off".

------

I propose a more _centralized_ approach to an order's tax-related calculations, where 
1. Each order-total module:
    1.  During its `process` method, returns the following values in its `output` array in addition to pre-existing fields:
       1.  Its _untaxed_ value.
       2. The `tax_class_id` associated with taxed element, mapping to an entry in the order's `tax_groups` array.  If the field is not returned or if its value is 0, the order-total's value is _presumed_ to be untaxed.
    2. No longer includes 'one-off' tax-calculation methods, e.g. `ot_group_pricing` and `ot_coupon`.
    3. If the module has dependencies on the _products_ in the order, the module uses the products _as recorded in the order_ for its calculations &hellip; **not** the products in the shopping-cart.
2. The order's `tax_groups` array is now indexed by a `tax_class_id` instead of a `tax_description`, supported by new order-class methods.  Those methods record the tax group's associated
	1. `tax_description`, as determined by the order-class.
	2. `tax_rate`, a currency-value in the range 0 to 100, as determined by the order-class.
	3. `product_total`, a currency-value that identifies the current _product_-related total for this tax-group. If a store displays prices with tax, this value is tax-inclusive.
	4. `additional_total`, a currency-value that identifies the current non-product total for this tax-group.  If a store displays prices with tax, this value is tax-inclusive.
3. All product-related deductions are recorded within the order's `products` array, preferably using a new order-class method, identifying
	1. The deduction `value`, either a percentage or a fixed currency-value.
	2. The deduction `type`, identifying whether the value is a percentage or fixed value.  Fixed values, for stores that display prices with tax, are tax-inclusive.
3. _All_ of the order's `tax_groups` are now recorded in the database, as multiple instances of the `ot_tax` order-total.  The order's tax display, under the control of `ot_tax` will either display the individual tax-rates or an aggregate value, under the control of _**Configuration->My Store->Show Split Tax Lines**_.

------

I'm still working through the specifics of the process-flow for this centralized approach, but wanted to get the basics out there for review before I dive any deeper.

**Comment by torvista** [comment]
Without bothering to tax (ha) my brain by trying to follow all that necessary explanation, and in the interests of repeating myself, does this cover all these eventualities?

https://wiki.dolibarr.org/index.php/VAT_calculation_and_rounding_rules

I mention this not as a proponent of Dolibarr but as I thought this is a good example of the various methods that have to be covered to be fully comprehensive. What a headache.

**Comment by lat9** [comment]
@torvista, very punny!  Thanks for adding that link (saves me having to hunt it up).

That's the goal.  If the tax-handling is centralized, then it will be easier for stores to customize any algorithm that might be required by their locality and/or shipping destinations.

**Comment by scottcwilson** [comment]
Is this a dup of #2121?

**Comment by drbyte** [comment]
> Is this a dup of #2121?

No; while related, it's different.

**Comment by scottcwilson** [comment]
@lat9 are you still keen on this? If not, please close; if so, please add label "Implement in Next Release" (or "Consider in Next Release") as you deem appropriate.

**Comment by scottcwilson** [comment]
Moving to Discussions.
